src/rpc/claimtrie.cpp

```diff
@@ -197,19 +209,32 @@ UniValue getvalueforname(const UniValue& params, bool fHelp)
         );
     LOCK(cs_main);
     std::string name = params[0].get_str();
-    CClaimValue claim;
+
+    int maxHeight = chainActive.Height();
+    if (params.size() > 1)
+        maxHeight -= params[1].get_int() - 1;
+
+    CCoinsViewCache coins(pcoinsTip);
+    CClaimTrieCache claimTrie(pclaimTrie);
+    if (maxHeight < chainActive.Height())
+        if (!RollBackTo(coins, claimTrie, chainActive[maxHeight]))
+            throw JSONRPCError(RPC_INTERNAL_ERROR, "Rollback failure");
+
     UniValue ret(UniValue::VOBJ);
-    if (!pclaimTrie->getInfoForName(name, claim))
-        return ret;
+    CClaimValue claim;
+    if (!claimTrie.getInfoForName(name, claim))
+        return ret; // they may have asked for a name that doesn't exist (which is not an error)
+
     std::string sValue;
-    if (!getValueForClaim(claim.outPoint, sValue))
+    if (!getValueForClaim(coins, claim.outPoint, sValue))
         return ret;
+
     ret.push_back(Pair("value", sValue));
     ret.push_back(Pair("claimId", claim.claimId.GetHex()));
     ret.push_back(Pair("txid", claim.outPoint.hash.GetHex()));
     ret.push_back(Pair("n", (int)claim.outPoint.n));
     ret.push_back(Pair("amount", claim.nAmount));
-    ret.push_back(Pair("effective amount", pclaimTrie->getEffectiveAmountForClaim(name, claim.claimId))); 
+    ret.push_back(Pair("effective amount", pclaimTrie->getEffectiveAmountForClaim(name, claim.claimId, maxHeight)));
```

Comment:
This particular line is not using data from the caches. It's hitting the live claimTrie itself. It appears that passing in the maxHeight would give the same result as rolling back the main trie.